rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -31,57 +31,75 @@ Status        | Proposed
-    the input data consistently.
-*   The complexity of adding new logical data representations (for example,
-    tf.SequenceExample) scales with the number of components.
+When a TFX component needs to invoke multiple TFX libraries, it may need to
+decode the data into one of the libraries’ in-memory representations,
+and translate that into the other library’s. For example, in the “Transform”
+component mentioned above:
+
+![alt_text](20191017-tfx-standardized-inputs/double-translation.png)
+
+Note that TFDV is invoked twice, with different data. Thus the translation needs
+to happen twice.
+
+This has created several issues:
+
+*   The translation is computationally expensive. In a real world set-up, such
+    translation could take as many CPU cycles as the core TFT and TFDV logic
+    takes in total.
+*   More such translation logic may need to be implemented to support expanding
+    TFX use cases -- imagine that TFMA needs to invoke TFDV to compute
+    statistics over slices of data identified by model evaluation.
+*   The complexity of adding new logical data representations scales with the
+    number of components. For example, to add SequenceExample support in TFX,
+    one may need to come up with an in-memory representation for each of the
+    components, to keep the consistency within the component.
+*   TFX library users whose data format is not supported natively by TFX would
+    have to implement the decoding logic for each of the libraries
+    they want to use.  For example, had TFX not supported the CSV format, a user
+    would have to implement
+    [one CSV decoder for TFT](https://github.com/tensorflow/transform/blob/master/tensorflow_transform/coders/csv_coder.py),
+    and [another CSV decoder for TFDV](https://github.com/tensorflow/data-validation/blob/master/tensorflow_data_validation/coders/csv_decoder.py)
+
+A common in-memory data representation would address the issues.
 
 ## The need for supporting new physical storage formats in TFX
 
-Two factors drive this need:
-
-*   TFX needs to offer users more choices of the storage format, e.g,
-    [Apache Parquet](https://parquet.apache.org/).
-*   TFX wants to be able to choose the optimal storage format based on user’s
-    workload, in a user-transparent manner. A unified I/O abstraction would make
-    it easier to support a new physical format in TFX, since one would not have
-    to understand every single TFX component in order to implement such support.
+Currently TFX (mostly) assumes tf.Example on TFRecord. However because TFX is a
+managed environment, it is desired that its choice of physical format of data is
+an implementation detail and is opaque to the components and the users. TFX
+would like to explore switching to a columnar physical format like Apache
+Parquet and it can be imagined that there will be a migration at some point.
+Such a migration must happen in either of the following two ways:
 
```

Comment:
At the end of the day, it's a migration of user pipelines, but by migration I mean both of

1. implementing support in libraries / components that need to read data.
2. changing user pipelines (trainer).

There is already such a migration happening internally (please see the internal design doc), which has followed the second approach because the first approach would require too many changes to all components, in addition to many user pipelines, if not all. And the second approach does not allow to realize all the efficiency benefits of a columnar storage format.

A similar migration will likely to happen in OSS, even if this proposal is rejected, but then we would have to make the same difficult choice.

The argument I was trying to make, is that if we have a abstract I/O layer, we wouldn't need to change any component / library, and we wouldn't need to change most user pipelines (as long as they use TFXIO.dataset()).